[CI] SearchableSnapshotDirectoryTests.testRecoveryStateIsKeptOpenAfterPreWarmFailures fails reproducibly on 7.x

**Build scan**:

https://gradle-enterprise.elastic.co/s/4uwypczipkc3g

**Repro line**:

./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.index.store.SearchableSnapshotDirectoryTests.testRecoveryStateIsKeptOpenAfterPreWarmFailures" \
  -Dtests.seed=F339D11550F542AE \
  -Dtests.security.manager=true \
  -Dtests.locale=zh \
  -Dtests.timezone=America/Martinique \
  -Druntime.java=8

**Reproduces locally?**: Yes

**Applicable branches**:

7.x

**Failure history**:

Some earlier ones on master on Aug 10th: https://gradle-enterprise.elastic.co/s/hwywdaugzxn6u

**Failure excerpt**:

java.lang.AssertionError: 
Expected: <FINALIZE>
     but: was <DONE>
	at __randomizedtesting.SeedInfo.seed([F339D11550F542AE:4E72F1FC5C3CCB98]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.index.store.SearchableSnapshotDirectoryTests.lambda$testRecoveryStateIsKeptOpenAfterPreWarmFailures$26(SearchableSnapshotDirectoryTests.java:779)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:959)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:932)
	at org.elasticsearch.index.store.SearchableSnapshotDirectoryTests.lambda$testRecoveryStateIsKeptOpenAfterPreWarmFailures$27(SearchableSnapshotDirectoryTests.java:779)
	at org.elasticsearch.index.store.SearchableSnapshotDirectoryTests.testDirectories(SearchableSnapshotDirectoryTests.java:620)
	at org.elasticsearch.index.store.SearchableSnapshotDirectoryTests.testRecoveryStateIsKeptOpenAfterPreWarmFailures(SearchableSnapshotDirectoryTests.java:778)

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java
@@ -775,7 +775,16 @@ public class SearchableSnapshotDirectoryTests extends ESTestCase {
         try {
             SearchableSnapshotRecoveryState recoveryState = createRecoveryState();
             testDirectories(true, true, recoveryState, Settings.EMPTY, (directory, snapshotDirectory) -> {
-                assertBusy(() -> assertThat(recoveryState.getStage(), equalTo(RecoveryState.Stage.FINALIZE)));
+                boolean areAllFilesReused = snapshotDirectory.snapshot()
+                    .indexFiles()
+                    .stream()
+                    .allMatch(fileInfo -> fileInfo.metadata().hashEqualsContents());
+                assertBusy(() -> {
+                    // When the number of indexed documents == 0, the index snapshot only contains the
+                    // commit file, meaning that the recovery won't fail in that case.
+                    RecoveryState.Stage expectedStage = areAllFilesReused ? RecoveryState.Stage.DONE : RecoveryState.Stage.FINALIZE;
+                    assertThat(recoveryState.getStage(), equalTo(expectedStage));
+                });
                 // All pre-warm tasks failed
                 assertThat(recoveryState.getIndex().recoveredBytes(), equalTo(0L));
             });

